stream-chat-android-core/src/main/java/io/getstream/chat/android/client/call/FlatMapCall.kt

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) 2014-2022 Stream.io Inc. All rights reserved.
+ *
+ * Licensed under the Stream License;
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://github.com/GetStream/stream-chat-android/blob/main/LICENSE
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.getstream.chat.android.client.call
+
+import io.getstream.chat.android.client.call.Call.Companion.callCanceledError
+import io.getstream.chat.android.client.utils.Result
+import io.getstream.chat.android.client.utils.flapMapCall
+import io.getstream.chat.android.client.utils.flatMapSuspend
+import io.getstream.chat.android.client.utils.map
+import io.getstream.chat.android.core.internal.coroutines.DispatcherProvider
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.withContext
+import java.util.concurrent.atomic.AtomicBoolean
+
+internal class FlatMapCall<T : Any, K : Any>(
+    private val call: Call<T>,
+    private val mapper: (T) -> Call<K>,
+) : Call<K> {
+
+    private val canceled = AtomicBoolean(false)
+
+    override fun cancel() {
+        canceled.set(true)
+        call.cancel()
+    }
+
+    override fun execute(): Result<K> = runBlocking { await() }
+
+    override fun enqueue(callback: Call.Callback<K>) {
+        call.enqueue { result ->
+            result.takeUnless { canceled.get() }
+                ?.flapMapCall(mapper)
```

Comment:
Here `flapMapCall ` will block `this` callback until

Reply:
🤔  But the code is inside `enqueue{}`, which is already async. The blocking will be the same as a normal `map` function.